### PR TITLE
Add TikTok to Andromeda

### PR DESCRIPTION
**Pull request by MarcosFleury**
*Adds TikTok Fetchers
** Needs Migrations from Pull Request #159 to work

* This Pull Requests has the commit history for adding TikTok to Andromeda

### CODE REVIEW

File: Andromeda.Commands/RunJobsCommand.cs [hunk 1]
```diff
@@ -7,16 +7,18 @@
...
 using Jobs.Fetcher.Twitter;
+using Jobs.Fetcher.TikTok;
 
 namespace Andromeda.Commands {
 
     public static class RunJobsCommand {
 
         public static JobsFactory[] Factories = new JobsFactory[] {
             new YouTubeFetchers(),
-            new AdWordsFetchers(),
```
Comment (moraesvic) [inline Andromeda.Commands/RunJobsCommand.cs:17]: Maybe disabling Adwords and reenabling Twitter should go in a separate PR -- but I'm just being pedantic

File: Jobs.Fetcher.TikTok/Helpers/ApiDataFetcher.cs [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+                var newEffectSticker = new EffectSticker() {
+                    Name = sticker["name"].ToString(),
+                    Id = sticker["ID"].ToString()
+                };
+                effectStickers.Add(newEffectSticker);
+            }
+            return effectStickers;
+        }
+
+        public static Author GetTikTokAuthorFromAuthorJSON(JToken authorJSON) {
```
Comment (moraesvic) [inline Jobs.Fetcher.TikTok/Helpers/ApiDataFetcher.cs:132]: When I'm coding, I often mix "JSON" and "Json", like you did here. Do you think we could establish a standard for the project?

File: Jobs.Fetcher.TikTok/Helpers/ApiDataFetcher.cs [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+                       PrivateAccount = authorJSON["privateAccount"].ToObject<bool>()
+            };
+        }
+
+        public static Author GetTikTokAuthorFromPostJSON(JToken postJSON) {
+            return new Author() {
+                       Id = postJSON["authorId"].ToString(),
+                       UniqueId = postJSON["author"].ToString(),
+                       Nickname = postJSON["nickname"].ToString()/*,
+                                                                    AvatarThumbnail = authorJSON["avatarThumb"].ToString(),
```
Comment (moraesvic) [inline Jobs.Fetcher.TikTok/Helpers/ApiDataFetcher.cs:159]: Will this get added later?

File: Jobs.Fetcher.TikTok/Helpers/ApiDataFetcher.cs [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+                       ValidityStart = DateTime.UtcNow,
+                       ValidityEnd = DateTime.MaxValue,
+                       AuthorId = author.Id,
+                       Author = author
+            };
+        }
+
+        public static Post GetTikTokPostFromJSON(JToken postJSON, Author author, Video video, Music music, List<string> challengeIds, List<string> tagIds, List<string> effectStickerIds) {
+            return new Post() {
+                       Id = postJSON["id"].ToString(),
```
Comment (moraesvic) [inline Jobs.Fetcher.TikTok/Helpers/ApiDataFetcher.cs:194]: I don't see any error handling. What would happen in case the JSON payload does not contain the fields we were expecting?

File: Jobs.Fetcher.TikTok/Helpers/DatabaseManager.cs [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+                        WHERE
+                            saved_time > @last_fetch :: timestamp without time zone AND
+                            account_name = @username
+                        ;");
+                    cmd.Parameters.AddWithValue("last_fetch", last_fetch.ToString("yyyy-MM-dd HH:mm:ss"));
+                    cmd.Parameters.AddWithValue("username", username);
+                    var payloadStrings = new List<string>();
+                    using (var reader = cmd.ExecuteReader()) {
+                        while (reader.Read()) {
+                            /*var log = new RowLog();
```
Comment (moraesvic) [inline Jobs.Fetcher.TikTok/Helpers/DatabaseManager.cs:36]:
```suggestion

```

File: Jobs.Fetcher.TikTok/Helpers/DatabaseManager.cs [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+                               log.AddInput("videos",
```
Comment (moraesvic) [inline Jobs.Fetcher.TikTok/Helpers/DatabaseManager.cs:37]:
```suggestion

```

File: Jobs.Fetcher.TikTok/Helpers/DatabaseManager.cs [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+                                        MutableEntityExtentions.AutoPK(reader.Prim<long>("id")));*/
```
Comment (moraesvic) [inline Jobs.Fetcher.TikTok/Helpers/DatabaseManager.cs:38]:
```suggestion

```

File: Jobs.Fetcher.TikTok/Helpers/DatabaseManager.cs [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+                        account_name
+                    WHERE
+                        account_name > @account_name
+                ");
+                cmd.Parameters.AddWithValue("account_name", account_name);
+                TikTokUsers tiktokUser = null;
+                using (var reader = cmd.ExecuteReader()) {
+                    if (reader.Read()) {
+                        tiktokUser = new TikTokUsers(){
+                            //UserId = reader.GetString(0),
```
Comment (moraesvic) [inline Jobs.Fetcher.TikTok/Helpers/DatabaseManager.cs:63]:
```suggestion

```

File: Jobs.Fetcher.TikTok/Helpers/DatabaseManager.cs [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+                            return returnValue;
+                        }
+                    }
+                    return false;
+                }
+        }
+
+        public static DateTime GetLastFetch(string authorId, DataLakeTikTokContext dbContext) {
+            var now = DateTime.UtcNow;
+            return dbContext.AuthorStats.Where(m => m.AuthorId == authorId && m.ValidityStart <= now && m.ValidityEnd > now)
```
Comment (moraesvic) [inline Jobs.Fetcher.TikTok/Helpers/DatabaseManager.cs:123]: I would also like to establish a standard for this timestamp range check. Are we going to check `ValidityEnd > now` or `ValidityEnd == DateTime.MaxValue`? I think the first one is the one normally used in the codebase

File: Jobs.Fetcher.TikTok/Helpers/DbReader.cs [hunk 10]
```diff
@@ -0,0 +1,16 @@
+using System;
```
Comment (moraesvic) [inline Jobs.Fetcher.TikTok/Helpers/DbReader.cs:1]: Is this file needed?

File: Jobs.Fetcher.TikTok/Helpers/DbWriter.cs [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+        }
+
+        public static void WriteMusic(Music newEntry, DataLakeTikTokContext dbContext, Logger logger) {
+            var oldEntry = dbContext.Music.Find(newEntry.Id);
+            Upsert<Music, DataLakeTikTokContext>(oldEntry, newEntry, dbContext, logger);
+            dbContext.SaveChanges();
+        }
+
+        public static void WritePost(Post newEntry, DataLakeTikTokContext dbContext, Logger logger) {
+            //var oldEntry = dbContext.Posts.Find(newEntry.Id);
```
Comment (moraesvic) [inline Jobs.Fetcher.TikTok/Helpers/DbWriter.cs:56]:
```suggestion

```

File: Jobs.Fetcher.TikTok/Jobs/PostsQuery.cs [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+
+        public PostsQuery(List<string> userIds): base(userIds) {}
+
+        public override List<string> Dependencies() {
+            return new List<string>() {};
+        }
+
+        public override void RunBody(string username) {
+            using (var dbContext = new DataLakeTikTokContext()) {
+                //Create a second loop for each TikTok client. Check with Erik and Victor if the Get option will receive an ID or username
```
Comment (moraesvic) [inline Jobs.Fetcher.TikTok/Jobs/PostsQuery.cs:21]: The interface for the scraper has been changed. Now we are using configuration files with JSON format (`jobs` directory tries to match the structure of `Credentials` directory in Andromeda), and the parameter read is the username

File: Jobs.Fetcher.TikTok/Jobs/PostsQuery.cs [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+                    DbWriter.WriteAuthorStats(newAuthorStats, dbContext, logger);
+
+                    var newPostStats = ApiDataFetcher.GetTikTokPostStatsJSON(post["stats"], newPost, newPost.CreateTime);
+                    DbWriter.WritePostStats(newPostStats, dbContext, logger);
+                }
+            }
+        }
+    }
+
+    public class ScrapperAccountAdd : AbstractTikTokFetcher {
```
Comment (moraesvic) [inline Jobs.Fetcher.TikTok/Jobs/PostsQuery.cs:87]: Typo: scrapper -> scraper

**Review by moraesvic — CHANGES_REQUESTED** [review]
The only thing that I think needs fixing is removing the table `account_names`, since now the scraper gets the users from credentials/jobs files and not from it. The rest are style corrections and questions to make sure we are on the same page.

File: Andromeda.Commands/RunJobsCommand.cs (hunk 1 above)
Comment (MarcosFleury) [reply Andromeda.Commands/RunJobsCommand.cs:17]: I agree it should, but I had already created so many different PRs that I decided to leave this one here

File: Jobs.Fetcher.TikTok/Helpers/ApiDataFetcher.cs (hunk 2 above)
Comment (MarcosFleury) [reply Jobs.Fetcher.TikTok/Helpers/ApiDataFetcher.cs:132]: I think having a standard is always a good idea. I just checked the rest of the project and it seems that 'Json' is more common, maybe because Newtonsoft uses this pattern. So I'll change everything to "Json"

File: Jobs.Fetcher.TikTok/Helpers/ApiDataFetcher.cs (hunk 3 above)
Comment (MarcosFleury) [reply Jobs.Fetcher.TikTok/Helpers/ApiDataFetcher.cs:159]: I forgot to remove this. This is the all the author data available when the Json file is incomplete

File: Jobs.Fetcher.TikTok/Helpers/DatabaseManager.cs (hunk 9 above)
Comment (MarcosFleury) [reply Jobs.Fetcher.TikTok/Helpers/DatabaseManager.cs:123]: Yeah, I think ```ValidityEnd > now``` is used more often, so I'll check any places where I might have used something different and change it

File: Jobs.Fetcher.TikTok/Jobs/PostsQuery.cs (hunk 12 above)
Comment (MarcosFleury) [reply Jobs.Fetcher.TikTok/Jobs/PostsQuery.cs:21]: I believe I had already fixed this, but I left this comment from long ago

**Review by moraesvic — APPROVED** [review]
